Make table `can-read?`/`can-write?` use `DataPermissions`

johnswanson opened the pull request.

qwef committed:
update table readwrite

qwef committed:
fix sandbox

johnswanson committed:
fix data permissions for table write

Also fix a test

johnswanson committed:
fix tests

First, make an enterprise/oss version of `current-user-can-write-table?` so
that we enforce data permissions correctly when `advanced-permissions`
is not enabled.

Second: I noticed something interesting in
`metabase-enterprise`.advanced-permissions.common-test. We
have `(u/ignore-exceptions (@#'perms/update-group-permissions! all-users-group-id graph))`.

I was curious about why we're ignoring exceptions, and one example is that we set data perms to:
`{db-id {:data {:native :write :schemas :block}}}`, and then assert that the user should be
able to access data.

But we're not REALLY setting those permissions, because if you remove
the ignore-exceptions you get:

```
   Invalid input: {53 {:data ["Invalid DB permissions: If you have write access
   for native queries, you must have data access to all schemas., got: {:native
   :write, :schemas {\"not_public\" :all}}"]}}
```

And the Permissions table ends up looking like this:

```
| /db/34/          |
| /download/db/34/ |
```

with no block permissions in sight!

It seems to me that we're not actually testing anything like what we say
we're testing here, so I removed the unnecessary and misleading tests.

johnswanson committed:
fix arg to `mi/can-read?`

johnswanson committed:
Fix sandboxed table API

This was a bit hairy, and I'm not confident I solved it in the best way.

The whole thing hinges on being able to add permission to a user. When a
user reads a sandboxed table through the API, then if they have the
necessary sandbox permissions, they just read the "normal" table API
with some additional permissions tacked on.

In the new DataPermissions world, this isn't quite as simple. How do we
tack on additional permissions?

I went through a couple options but ended up deciding that we could do
with something dead simple that just satisfies the need *here*, and
worry about other cases where we need to tack on additional permissions
later (if they exist - I'm not sure!).

So a new macro, `data-perms/with-additional-table-permission`, just adds
a single permission for a given db and table (stored in a dynamic var).
When we read table permission via
`data-perms/table-permission-for-user`, we read permission from the
database *and* from the dynamic var - coalescing like normal, which
means that in general the least restrictive permission "wins."

This doesn't affect any other functions besides
`table-permission-for-user` - so e.g. it doesn't impact whether all
tables have the same permission value when we're setting default
permissions.

I'm a little sad about this change, because it means that we're not
storing permissions in the database. Other than admin users, it's the
first time we're stepping away from permissions just being database
rows, and it feels sad to lose that transparency. But, I couldn't figure
out a way to do this kind of thing (tacking on an additional permission)
while maintaining that simplicity.

johnswanson committed:
fix with-gtaps!

I wasn't granting permissions on the table. Oops.

johnswanson committed:
fix another test

johnswanson committed:
Fix schema-level permissions

The old check for whether a user could read a given schema was:

```
   (or
    (perms/set-has-partial-permissions? @api/*current-user-permissions-set*
                                        (perms/data-perms-path database-id schema-name))
    (perms/set-has-full-permissions? @api/*current-user-permissions-set*
                                     (perms/data-model-write-perms-path database-id schema-name))))
```

I misinterpreted this. If a user has unrestricted access to one table in
a schema but not another table, they *can* read the schema (because
`/db/1/schema/foo/table/1` *is* a partial permission for
`/db/1/schema/foo`).

But, on the other hand, when we check whether a user can upload to a
given schema, we check that they have *full* schema permissions. So we
need two different ways to check schema permissions: "full" schema
permissions returns the most *restrictive* permission, while another
check returns the most *permissive* permission.

johnswanson committed:
fix another test

johnswanson committed:
Fix `call-count` test

We're doing more DB calls as we get permissions right now - that's ok.

johnswanson committed:
Fix more tests

johnswanson committed:
fix more tests!

noahmoss commented on enterprise/backend/src/metabase_enterprise/sandbox/api/table.clj:
```diff
@@ -28,14 +27,15 @@
 (mu/defn only-sandboxed-perms? :- :boolean
-  "Returns true if the user has only segemented and not full table permissions. If the user has full table permissions
-  we wouldn't want to apply this segment filtering."
+  "Returns true if the user has sandboxed permissions. If a sandbox policy exists, it overrides existing permission on
+  the table."
```
There's a small distinction here between a sandboxing policy existing, and sandboxing actually being enforced for the user. Sandboxing is only enforced if the user isn't in any _other_ group which grants full access to the table.

It looks like the original function was checking whether sandboxing is enforced, while the new one is just checking whether any sandboxing policy exists.

I'm not sure if this distinction matters much for this namespace but it would definitely matter for the usages in `metabase-enterprise.sandbox.models.params.field-values` since it computes a different hash key for cached field values based on whether sandboxing is enforced or not.

To match the previous behavior you'd need to use `metabase-enterprise.sandbox.api.util/enforced-sandboxes`, or at least do something similar here.

johnswanson replied:
Thanks for noticing this and calling it out. I think I mentioned this in slack, but I had been thinking that our move from "sandbox is its own permission" to "sandbox is unrestricted perm + a GTAP" meant that we didn't need to check for unrestricted perms anymore, but as you pointed out, that's wrong - we need to check for unrestricted perms from _other_ groups.

I ended up modifying the `sandbox.api.util` namespace to add an `enforced-sandboxes-for` function. It takes a user ID and returns the sandboxes that will be enforced for that user. Then this function checks to see whether those enforced sandboxes include one for this table.

noahmoss replied:
Cool, makes sense!

noahmoss commented on enterprise/backend/src/metabase_enterprise/sandbox/api/table.clj:
```diff
@@ -63,21 +63,18 @@
-        (thunk))
+      (maybe-filter-fields
+       table
+       (data-perms/with-additional-table-permission :perms/data-access (:db_id table) (u/the-id table) :unrestricted
```
Nice, this is super useful

noahmoss commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/common_test.clj:
```diff
@@ -789,17 +795,12 @@
-                       #"You don't have permissions to do that\."
-                       (append-csv!))))))))))))
+          (testing "With blocked perms it should fail"
+            (with-all-users-data-perms! {db-id {:data {:native :none,  :schemas :block}}}
```
nit
```suggestion
            (with-all-users-data-perms! {db-id {:data {:native :none, :schemas :block}}}
```

johnswanson replied:
fixed in a rebase, thanks!

noahmoss reviewed (changes requested):
Looks great! Just think there's one small correctness point about the sandboxing change.

johnswanson committed:
Update `enforced-sandboxes` to use DataPermissions

johnswanson committed:
remove broken tests

`with-all-users-data-perms!` used to avoid throwing exceptions if we set
invalid permissions. I changed this to throw the exceptions, which I
think is more reasonable: we weren't actually setting the permissions we
said we were setting. For example, a test that declares that we're setting
permissions to `{:native :write :schemas :block}` would *actually* be
setting permissions to block, without any native query permissions.

johnswanson committed:
oops, make sandbox hydration more efficient

forgot about `t2/hydrate` operating on multiple instances at once

johnswanson committed:
remove single-item doseq
